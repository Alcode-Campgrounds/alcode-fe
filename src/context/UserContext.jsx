import { useContext, useState, createContext, useEffect } from 'react';
import { getCurrentUser } from '../utils/AuthFetch';

const UserContext = createContext();

const UserProvider = ({ children }) => {
    const [user, setUser] = useState(null)
    const [loading, setLoading] = useState(true)

    useEffect(() => {
        getCurrentUser()
            .then(user => setUser(user))
            .catch(error => console.log(error))
            .finally(() => setLoading(false))
    }, [])

    return (
        <UserContext.Provider value={{ user, setUser, loading }}>
            {children}
        </UserContext.Provider>
    )
}

const useLoadingUser = () => {
    const { loading } = useContext(UserContext)
    return loading
}

const useUser = () => {
    const context = useContext(UserContext);
    if (context === undefined) {
        throw new Error('useUser hook must be called inside of a GuessContext Provider');
    }
    return context;
}

//placed here for comparison
const useAuth = () => {
    // useUser is its own custom hook and not a context. 
    // This useAuth isn't doing anything useUser isn't. It's returning the same context
    const context = useContext(useUser)

    if (context === undefined) {
        throw new Error('useAuth must be used within a provideAuth context');
    }

    return context;
}

export { UserProvider, useUser, useLoadingUser }
