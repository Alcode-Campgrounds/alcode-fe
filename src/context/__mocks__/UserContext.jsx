import { useContext, useState, createContext, useEffect } from 'react';
// import { getCurrentUser } from '../utils/AuthFetch';

const UserContext = createContext();

const UserProvider = ({children}) => {
    const [user, setUser] = useState(null)
    const [loading, setLoading] = useState(true)
    
    useEffect(() => { 
        setUser({name: 'fake-user', email: 'fake@email.com'})
        setLoading(false)
            // getCurrentUser()
            //     .then(user => setUser(user))
            //     .catch(error => console.log(error))
            //     .finally(()=>setLoading(false)) 
    }, [])

    return (
        <UserContext.Provider value={{user, setUser, loading}}>
            {children}
        </UserContext.Provider>
    )
}
const useLoadingUser = () => {
    const { loading } = useContext(UserContext)
    return loading
}
const useUser = () => { 
    const context = useContext(UserContext);
    if (context === undefined){
        throw new Error('useUser hook must be called inside of a GuessContext Provider');
    }
    return context;
}

export { UserProvider, useUser, useLoadingUser}