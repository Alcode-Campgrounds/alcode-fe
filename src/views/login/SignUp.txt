// import React from 'react'
// import { create }from 'react-test-renderer'
// import { UserProvider } from '../../context/UserContext'
// import SignUp from './SignUp'
// import { MemoryRouter as Router, Routes, Route } from 'react-router-dom'
 

 



// describe('is it here?', () => {
//     it('did it render?', () => {
    
//     const wrapper = create(
//     <UserProvider>
//         <Router initialEntries={["/signup"]}>
//             <Routes>
//                 <Route path="signup" element={<SignUp />} />
//             </Routes>
//         </Router>
//     </UserProvider>
//     )

//         expect(wrapper).toMatchSnapshot()
//     })
// })

