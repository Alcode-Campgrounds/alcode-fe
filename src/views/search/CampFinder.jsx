import React, { useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";
import { fetchAllFacilities, fetchStateFacility } from "../../utils/campFetch";
import { getStorage, setStorage } from "../../utils/localStorage";
import States from "./States";
// import { Loader } from "../../components/loader/"

export default function CampFinder() {
  const [state, setState] = useState('ALL');
  const [loading, setLoading] = useState(true)
  console.log(loading)
  const navigate = useNavigate();

  useEffect(() => {
    async function getCampgrounds() {
      const existingStorage = getStorage('ALL');
      if (existingStorage.length === 0) {
        const facilities = await fetchAllFacilities();
        setLoading(false);
        setStorage('ALL', facilities);
      }
    }
    getCampgrounds();
  }, []);

  const handleStateSubmit = async (e) => {
    e.preventDefault();
    if (state === 'ALL') {
      navigate(`/campmap/${state}`)    
    } else {
      const existingStateStorage = getStorage(state);
      if (existingStateStorage.length === 0) {
        const stateFacility = await fetchStateFacility(state);
        setStorage(state, stateFacility);
        navigate(`/campmap/${state}`)   
      } else {
        navigate(`/campmap/${state}`) 
      }
    }
  };

  const handleStateChange = (e) => {
    setState(e.target.value);
  }

  return (
    <>

      <h3 className='search-heading'>Search for all campgrounds by State</h3>
      <form onSubmit={handleStateSubmit} className='search-form'>
        <States setState={handleStateChange} />
        <button type="submit" aria-label="Search" className='search-btn'>
          Search
        </button>
      </form>
    
    </>
  );
}
     
      
