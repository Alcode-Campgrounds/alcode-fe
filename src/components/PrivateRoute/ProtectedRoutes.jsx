import React from "react";
import {  Navigate, Outlet} from 'react-router-dom';
import { useUser } from "../../context/UserContext";



export default function ProtectedRoutes(){
  // const useAuth = () => {
    const {user, loading } = useUser();
  //   if (!loading) return user.email
  // }
  // const isAuth = useAuth();
  // console.log('Auth', isAuth)
  console.log('PRIVATE', user);
  if (!loading && user){
    return <Outlet />
  } else {
    return <Navigate to='/signin'/>
  }
  // return user ? <Outlet/> : <Navigate to='/signin'/>
}

