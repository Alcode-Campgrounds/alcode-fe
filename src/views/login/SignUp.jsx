import React from 'react';
import { useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { useUser } from '../../context/UserContext';
import { existingUserSignIn, newUserSignUp } from '../../utils/AuthFetch';

export default function SignUp({ hasUser = false }) {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [name, setName] = useState('');
  const [error, setError] = useState('');
  // const [gooBtn, setGooBtn] = useState(false)
  const { setUser } = useUser();

  let navigate = useNavigate();

  const handleSubmitSignUp = async (e) => {
    e.preventDefault();

    if (hasUser) {
      const userData = await existingUserSignIn(email, password);
      if (userData.email) {
        setUser({ name: userData.name, email: userData.email });
        navigate('/search');
      } else {
        setError('Wrong email/password');
      }
    } else {
      try {
        const userSignUp = await newUserSignUp(name, email, password);
        console.log(userSignUp.status);
        if (userSignUp) {
          setUser({ name, email });
          navigate('/search');
        }
      } catch (error) {
        setError(error.message);
      }
    }
  };

  return (
    <form className='form-main' onSubmit={handleSubmitSignUp}>
      <fieldset>
        <legend className='legend'>{hasUser ? 'SignIn' : 'SignUp'}</legend>
      
          <a className='google-btn' href={process.env.REACT_APP_OAUTH}>
          Google
          </a>
        <label htmlFor='email-input'>
          <input
            id='email-input'
            required
            value={email}
            type='text'
            placeholder='email'
            className='input-A'
            onChange={({ target }) => {
              setEmail(target.value);
            }}
          />
        </label>
        <label htmlFor='password-input'>
          <input
            id='password-input'
            required
            value={password}
            type='text'
            placeholder='password'
            className='input-B'
            onChange={({ target }) => {
              setPassword(target.value);
            }}
          />
        </label>

        {!hasUser && (
          <label htmlFor='name-input'>
            <input
              id='name-input'
              required
              value={name}
              type='text'
              placeholder='name'
              className='input-C'
              onChange={({ target }) => {
                setName(target.value);
              }}
            />
          </label>
        )}
        <br />
        <button type='submit' className='primary-btn'>
          {hasUser ? 'SignIn' : 'SignUp'}
        </button>
        <p>{error}</p>
      </fieldset>
    </form>
  );
}
