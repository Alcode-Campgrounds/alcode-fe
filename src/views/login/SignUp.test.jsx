import {render, screen} from '@testing-library/react';
import { UserProvider } from '../../context/UserContext';
import SignUp from './SignUp';
import {MemoryRouter, Route, Routes} from 'react-router-dom';
import CampFinder from '../search/CampFinder';
import userEvent from '@testing-library/user-event';
import ProtectedRoutes from '../../components/PrivateRoute/ProtectedRoutes';
import { setupServer } from 'msw/node';
import { rest } from 'msw';

const fakeUser = {
    name: 'fake-user',
    email: 'fakeUser@email.com',
    password: 'fakepassword'
}
const worker = setupServer(
    rest.post(process.env.REACT_APP_BACKEND_URL, (req, res, ctx) => {
        return res(
            ctx.json(fakeUser)
        )
    })
)
beforeAll(() => {
    worker.listen({ onUnhandledRequest: "bypass" })
});
afterAll(() => {
    worker.close()
});

it('should render a new user signing up', async () => {
    render(
        <UserProvider>
            <MemoryRouter initialEntries={['/signup']}>
                <Routes>
                    <Route path="/signup" element={<SignUp />} />
                    <Route element={<ProtectedRoutes />}>
                        <Route path="/search" element={<CampFinder />} />
                    </Route>
                </Routes>
            </MemoryRouter>
        </UserProvider>
    )
    const email = await screen.findByPlaceholderText('email');
    const password = await screen.findByPlaceholderText('password');
    const name = await screen.findByPlaceholderText('name');
    const signupButton = await screen.findByText('SignUp');
    
    expect(email).toBeInTheDocument();
    expect(password).toBeInTheDocument();
    expect(name).toBeInTheDocument();
    expect(signupButton).toBeInTheDocument();

    userEvent.type(email, 'fakeUser@email.com');
    userEvent.type(password, 'fakepassword');
    userEvent.type(name, 'fake-user');
    userEvent.click(signupButton)
    
    const search = await screen.findByText(/Search for all campgrounds by State/i);
    expect(search).toBeInTheDocument();
});
