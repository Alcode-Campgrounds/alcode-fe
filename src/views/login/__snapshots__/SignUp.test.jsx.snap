// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`is it here? did it render? 1`] = `
<form
  className="form-main"
  onSubmit={[Function]}
>
  <fieldset>
    <legend>
      SignUp
    </legend>
    <a
      href="https://nameless-taiga-88760.herokuapp.com/api/alcode/auth/login"
    >
      Sign in with your Google account
    </a>
    <label
      htmlFor="email-input"
    >
      <input
        className="input"
        id="email-input"
        onChange={[Function]}
        placeholder="email"
        required={true}
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="password-input"
    >
      <input
        className="input"
        id="password-input"
        onChange={[Function]}
        placeholder="password"
        required={true}
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="name-input"
    >
      <input
        className="input"
        id="name-input"
        onChange={[Function]}
        placeholder="name"
        required={true}
        type="text"
        value=""
      />
    </label>
    <button
      className="btn-main"
      type="submit"
    >
      SignUp
    </button>
    <p>
      
    </p>
  </fieldset>
</form>
`;
